prov/efa/src/efa_mr.c

```diff
@@ -227,7 +221,32 @@ static int efa_mr_hmem_setup(struct efa_mr *efa_mr,
 		efa_mr->needs_sync = true;
 		efa_mr->peer.device.cuda = attr->device.cuda;
 
-		if (cuda_is_gdrcopy_enabled()) {
+		if ((flags & FI_MR_DMABUF) != 0) {
+			if (!cuda_is_dmabuf_supported()) {
```

Comment:
I'm happy to go with whatever you'd prefer.

Part of the difficulty on our end with this feature is that we don't have a great way to get a direct answer from libfabric around dmabuf support ahead of time -- all we can do is probe for 1.20 + FI_HMEM and then proceed optimistically. (Well, we do have the option to query the allocation granularity, and then use that to create a mempool, and then reserve an address range, and then map the created mempool on that address range, and export that address range, and then initialize a provider, and then attempt a registration, and then tear it all down... but this becomes fairly intrusive for us). Long term, I think the API as it exists today (where dmabuf support is implied for FI_HMEM beyond some api version) is the ideal. But today there are integration pains.

Example of the pain: There exist real systems (ie: p5 + AL2 on the 5.10 kernel) where dmabuf export through cuda works, but the ib ioctls don't exist on the kernel side until linux 5.12, meaning `ib_reg_mr_dmabuf` will statically fail. In those cases, it's not wrong for `CU_DEVICE_ATTRIBUTE_DMA_BUF_SUPPORTED` to return true, because it is a valid exporter for importers like mesa/v4l/whatever, but it's insufficient for our purposes. Whose responsibility should it be to check for kernel import support? In practice, ofi-nccl added a ~5 line function to call uname during init so that we can test the major/minor version. But IMO this feels like an abstraction has leaked and I would have preferred to be able to assume that libfabric is validating this for us. Prior to adding the check, libfabric didn't do anything to stop it from getting to the point of making a call to `ib_reg_mr_dmabuf`. (I think this fallback logic [in efa_domain_hmem_info_check_p2p_support_cuda](https://github.com/ofiwg/libfabric/blob/v1.22.x/prov/efa/src/efa_hmem.c#L154) is responsible; On the impacted 5.10 systems, where `CU_DEVICE_ATTRIBUTE_DMA_BUF_SUPPORTED` returns true, `ibv_reg_dmabuf_mr` will fail but the second call to `ibv_reg_mr` can succeed, meaning `p2p_supported_by_device` remains hot.)

The reason I'm bringing this case up in the context of an assert here is just to mention that there are a multitude of reasons that nccl + ofi-nccl can believe dmabuf support is available whereas you might not, and while I think we're covered today, if and when that happens we will end up hitting this assert. This is the first opportunity that we get to receive a strong yes/no signal from libfabric that there's agreement that dmabuf requirements are met... When we disagree, I'd prefer to get control returned so that we can print an explanation of what probably happened and provide instructions on how to force the non-dmabuf path.